File: VortexVise/Hook.cpp
#include "Hook.h"

void Hook::Simulate(Player& player, Map& map, const float& gravity, const float& deltaTime, Input& input)
{
	if (input.CancelHook && m_isHookAttached) {
		m_isHookReleased = false;
		m_isHookAttached = false;
		m_velocity = { 0,0 };

	}
	if (!m_pressingHookKey && input.Hook) {
		// start Hook shoot
		m_isHookReleased = true;
		m_isHookAttached = false;
		m_position = player.GetPlayerCenterPosition();
		m_collision.x = m_position.x;
		m_collision.y = m_position.y;

		// Reset velocity
		m_velocity = { 0,0 };

		// Get hook direction
		if (input.Left && input.Down){
			// ↙ 
			m_velocity.x -= m_hookShootForce;
			m_velocity.y += m_hookShootForce;
		}
		else if (input.Right && input.Down){
			// ↘
			m_velocity.x += m_hookShootForce;
			m_velocity.y += m_hookShootForce;
		}
		else if (input.Down){
			// ↓
			m_velocity.y += m_hookShootForce;
		}
		else if (input.Left && input.Up) {
			// ↖
			m_velocity.x -= m_hookShootForce * 0.6;
			m_velocity.y -= m_hookShootForce;
		}
		else if (input.Right && input.Up) {
			// ↗
			m_velocity.x += m_hookShootForce * 0.6;
			m_velocity.y -= m_hookShootForce;
		}
		else if (input.Left) {
			// ↖
			m_velocity.x -= m_hookShootForce;
			m_velocity.y -= m_hookShootForce;
		}
		else if (input.Right) {
			// ↗
			m_velocity.x += m_hookShootForce;
			m_velocity.y -= m_hookShootForce;
		}
		else if (input.Up) {
			// ↑
			m_velocity.y -= m_hookShootForce;
		}
		else {
			// This will use the player direction
			if (player.IsLookingRight()) {
				m_velocity.x += m_hookShootForce;
				m_velocity.y -= m_hookShootForce;
			}
			else {
				m_velocity.x -= m_hookShootForce;
				m_velocity.y -= m_hookShootForce;

			}
		}
	}
	else if ((m_pressingHookKey && !input.Hook)) {
		// Hook retracted
		m_isHookReleased = false;
		m_isHookAttached = false;
		m_velocity = { 0,0 };
	}
	else if (!m_isHookAttached) {
		// Shooting the hook
		m_position = { m_position.x + m_velocity.x * deltaTime,m_position.y + m_velocity.y * deltaTime };
		m_position.y += gravity * 0.5 * deltaTime;
		m_collision.x = m_position.x;
		m_collision.y = m_position.y;

	}
	else if (m_isHookAttached) {
		// Should pull player here
		Vector2 direction = Utils::GetVector2Direction(player.GetPlayerCenterPosition(), m_position);

		float distance = Vector2Distance(m_position, player.GetPosition());

		// TODO: implement this crap here
		//if((m_hookPullOffset > m_originalPullOffset && _offsetChanger < 0) || (_hookPullOffset < _originalPullOffset * 6 && _offsetChanger > 0))
		//{
		//_hookPullOffset += _offsetChanger * Time.deltaTime * 10;

		//if(_soundTimer == 0)  // This is to not spam the audio 
		//{
		//GameObject.FindWithTag("AudioSystem").GetComponent<AudioSystem>().PlayElastic();
		//_soundTimer += Time.deltaTime;
		//}


		if (distance > m_hookPullOffset)
		{
			Vector2 velocity = Vector2Scale(direction, m_hookPullForce);
			player.AddVelocity(velocity, deltaTime);
		}
	}
	m_pressingHookKey = input.Hook;

	if (m_isHookReleased) {
		for (const auto& collision : map.GetCollisions()) {
			if (CheckCollisionRecs(m_collision, collision)) {
				m_isHookAttached = true;
			}
		}
	}
}

void Hook::Draw(Player& const player)
{
	if (m_isHookReleased) {
		DrawLineEx(player.GetPlayerCenterPosition(), { m_position.x + 8, m_position.y + 8 }, 2, { 159,79,0,255 });
		DrawTexture(m_texture, m_position.x - 8, m_position.y - 10, WHITE);

		if (Utils::Debug())
			DrawRectangleRec(m_collision, GREEN); // Debug
	}

}

File: VortexVise/GameObjects/Hook.cs
using Raylib_cs;
using System.Numerics;
using VortexVise.Models;
using VortexVise.Networking;
using VortexVise.Utilities;

namespace VortexVise.GameObjects;

public class Hook
{
    Vector2 _position = new(0, 0);
    Vector2 _velocity = new(0, 0);
    Rectangle _collision = new(0, 0, 16, 16);
    Texture2D _texture = Raylib.LoadTexture("Resources/Sprites/GFX/hook_head.png");
    float _hookPullForce = 2500;
    float _hookPullOffset = 50;
    float _hookShootForce = 2000;
    float _hookSizeLimit = 200;
    bool _isHookAttached = false;
    bool _isHookReleased = false;
    float _hookTimeout = 0.2f;
    bool _pressingHookKey = false;
    private HookState GetState()
    {
        var state = new HookState();
        state.Position = _position;
        state.Velocity = _velocity;
        state.Collision = _collision;
        state.IsHookAttached = _isHookAttached;
        state.IsHookReleased = _isHookReleased;
        state.IsPressingHookKey = _pressingHookKey;
        return state;
    }
    public HookState Simulate(Player player, Map map, float gravity, float deltaTime, Input input)
    {
        var state = GetState();
        if (input.CancelHook && state.IsHookAttached)
        {
            state.IsHookReleased = false;
            state.IsHookAttached = false;
            state.Velocity = new(0, 0);

        }
        if (!state.IsPressingHookKey && input.Hook)
        {
            // start Hook shoot
            state.IsHookReleased = true;
            state.IsHookAttached = false;
            state.Position = player.GetPlayerCenterPosition();
            state.Collision.X = state.Position.X;
            state.Collision.Y = state.Position.Y;

            // Reset velocity
            state.Velocity = new(0, 0);

            // Get hook direction
            if (input.Left && input.Down)
            {
                // ↙ 
                state.Velocity.X -= _hookShootForce;
                state.Velocity.Y += _hookShootForce;
            }
            else if (input.Right && input.Down)
            {
                // ↘
                state.Velocity.X += _hookShootForce;
                state.Velocity.Y += _hookShootForce;
            }
            else if (input.Down)
            {
                // ↓
                state.Velocity.Y += _hookShootForce;
            }
            else if (input.Left && input.Up)
            {
                // ↖
                state.Velocity.X -= _hookShootForce * 0.6f;
                state.Velocity.Y -= _hookShootForce;
            }
            else if (input.Right && input.Up)
            {
                // ↗
                state.Velocity.X += _hookShootForce * 0.6f;
                state.Velocity.Y -= _hookShootForce;
            }
            else if (input.Left)
            {
                // ↖
                state.Velocity.X -= _hookShootForce;
                state.Velocity.Y -= _hookShootForce;
            }
            else if (input.Right)
            {
                // ↗
                state.Velocity.X += _hookShootForce;
                state.Velocity.Y -= _hookShootForce;
            }
            else if (input.Up)
            {
                // ↑
                state.Velocity.Y -= _hookShootForce;
            }
            else
            {
                // This will use the player direction
                if (player.IsLookingRight())
                {
                    state.Velocity.X += _hookShootForce;
                    state.Velocity.Y -= _hookShootForce;
                }
                else
                {
                    state.Velocity.X -= _hookShootForce;
                    state.Velocity.Y -= _hookShootForce;

                }
            }
        }

        else if ((state.IsPressingHookKey && !input.Hook))
        {
            // Hook retracted
            state.IsHookReleased = false;
            state.IsHookAttached = false;
            state.Velocity = new(0, 0);
        }
        else if (!state.IsHookAttached)
        {
            // Shooting the hook
            state.Position = new(state.Position.X + state.Velocity.X * deltaTime, state.Position.Y + state.Velocity.Y * deltaTime);
            state.Position.Y += gravity * 0.5f * deltaTime;
            state.Collision.X = state.Position.X;
            state.Collision.Y = state.Position.Y;

        }
        else if (state.IsHookAttached)
        {
            // Should pull player here
            Vector2 direction = Utils.GetVector2Direction(player.GetPlayerCenterPosition(), state.Position);

            float distance = Raymath.Vector2Distance(state.Position, player.GetPosition());

            // TODO: implement this crap here
            //if((_hookPullOffset > _originalPullOffset && _offsetChanger < 0) || (_hookPullOffset < _originalPullOffset * 6 && _offsetChanger > 0))
            //{
            //_hookPullOffset += _offsetChanger * Time.deltaTime * 10;

            //if(_soundTimer == 0)  // This is to not spam the audio 
            //{
            //GameObject.FindWithTag("AudioSystem").GetComponent<AudioSystem>().PlayElastic();
            //_soundTimer += Time.deltaTime;
            //}


            if (distance > _hookPullOffset)
            {
                Vector2 velocity = Raymath.Vector2Scale(direction, _hookPullForce);
                player.AddVelocity(velocity, deltaTime);
            }
        }
        state.IsPressingHookKey = input.Hook;

        if (state.IsHookReleased)
        {
            foreach (var collision in map.GetCollisions())
            {
                if (Raylib.CheckCollisionRecs(state.Collision, collision))
                {
                    state.IsHookAttached = true;
                }
            }
        }
        return state;
    }

    public void Draw(Player player)
    {
        if (_isHookReleased)
        {
            Raylib.DrawLineEx(player.GetPlayerCenterPosition(), new Vector2( _position.X + 8, _position.Y + 8 ), 2, new Color( 159,79,0,255 ));
            Raylib.DrawTexture(_texture, (int)_position.X - 8, (int)_position.Y - 10, Color.White);

            if (Utils.Debug())
                Raylib.DrawRectangleRec(_collision, Color.Green); // Debug
        }



    }
    public void ApplyState(HookState state)
    {
        _position = state.Position;
        _velocity = state.Velocity;
        _collision = state.Collision;
        _isHookAttached = state.IsHookAttached;
        _isHookReleased = state.IsHookReleased;
        _pressingHookKey = state.IsPressingHookKey;
    }
}
